pkg/reconciler/pipeline/dag/dag_test.go

```diff
@@ -669,3 +669,59 @@ func TestBuild_ConditionsParamsFromTaskResults(t *testing.T) {
 	}
 	assertSameDAG(t, expectedDAG, g)
 }
+
+// This test make sure we detect cycle when it exists.
+// This means we "visit" "a" twice, from two different paths.
+// And one of the paths is creating a cycle with "e -> a".
+//   a
+//  / \
+// b   c
+//  \ /
+//   d
+//  / \
+// e   f
+//     |
+//     g
+func TestBuild_InvalidDAGWithCycle(t *testing.T) {
+	a := v1beta1.PipelineTask{Name: "a", RunAfter: []string{"e"}}
+	bDependsOnA := v1beta1.PipelineTask{
+		Name: "b",
+		Resources: &v1beta1.PipelineTaskResources{
+			Inputs: []v1beta1.PipelineTaskInputResource{{From: []string{"a"}}},
+		},
+	}
+	cRunsAfterA := v1beta1.PipelineTask{
+		Name:     "c",
+		RunAfter: []string{"a"},
+	}
+	dDependsOnBAndC := v1beta1.PipelineTask{
+		Name: "f",
+		Resources: &v1beta1.PipelineTaskResources{
+			Inputs: []v1beta1.PipelineTaskInputResource{{From: []string{"b", "c"}}},
+		},
+	}
+	eRunsAfterD := v1beta1.PipelineTask{
+		Name:     "e",
+		RunAfter: []string{"d"},
+	}
+	fRunsAfterD := v1beta1.PipelineTask{
+		Name:     "f",
+		RunAfter: []string{"d"},
+	}
+	gDependsOnF := v1beta1.PipelineTask{
+		Name: "g",
+		Resources: &v1beta1.PipelineTaskResources{
+			Inputs: []v1beta1.PipelineTaskInputResource{{From: []string{"f"}}},
+		},
+	}
+
+	p := &v1beta1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{Name: "invalid-dag-with-cycle"},
+		Spec: v1beta1.PipelineSpec{
+			Tasks: v1beta1.PipelineTaskList{a, bDependsOnA, cRunsAfterA, dDependsOnBAndC, eRunsAfterD, fRunsAfterD, gDependsOnF},
+		},
+	}
+	if _, err := dag.Build(v1beta1.PipelineTaskList(p.Spec.Tasks)); err == nil {
```

Comment:
Hey @sbwsg, there are multiple paths possible leading to a cycle. It all depends on the order in which dependency map is processed. 

This test can fail with one of these paths:

```
	if err == nil || !(strings.Contains(err.Error(), "e -> a -> b -> d -> e") ||
		strings.Contains(err.Error(), "a -> c -> b -> d -> e -> a") ||
		strings.Contains(err.Error(), "d -> e -> a -> b -> d")) {
		t.Errorf("expected cycle error but got %v", err)
	}
```

I can replace the existing error checking with the above check if we decide to keep this test.

Reply:
Running the unit test multiple times with `count=1` leads to many different combination of paths including:

```
		"a -> c -> b -> d -> e -> a",
		"a -> b -> c -> d -> e -> a",
		"a -> b -> d -> e -> a",
		"a -> c -> d -> e -> a",
		"b -> d -> e -> a -> b",
		"c -> d -> e -> a -> c",
```

I walked through the graph one more time and there are much more combinations possible. I think its fine to leave it as just `cycle detected` instead of looking for a specific path in the unit test.

Reply:
Great point, the order of traversal isn't stable across runs because the nodes are stored in a hash map.  `cycle detected` makes sense 👍